paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc

```diff
@@ -385,6 +547,138 @@ class DNNLMatMulKernel : public framework::OpKernel<T> {
     ExecuteMatMul<T, T>(ctx);
   }
 };
+
+template <typename T>
+class MatMulGradMKLDNNKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const ExecutionContext& ctx) const override {
+    if (ctx.HasAttr("head_number")) {
+      PADDLE_ENFORCE_EQ(
+          ctx.Attr<int>("head_number"), 1,
+          platform::errors::Unimplemented(
+              "DNNL matmul doesn't support multiple heads. Expected "
+              "head_number=1. But received `head_number` is %d",
+              ctx.Attr<int>("head_number")));
+    }
+    RunKernel<T>(ctx);
+  }
+
+ private:
+  void ExecuteMatMulGrad(const ExecutionContext& ctx,
+                         const MKLDNNDeviceContext& dev_ctx,
+                         const mkldnn::engine& engine, Tensor* x, bool trans_x,
+                         bool is_fold_init_dims_x, Tensor* y, bool trans_y,
+                         bool is_fold_init_dims_y, Tensor* out,
+                         int execution_number) const {
+    // gradient is calculated in a different way when broadcasting is used
+    bool need_combine = (x->dims().size() == 3 || y->dims().size() == 3) &&
+                        out->dims().size() == 2;
+
+    Tensor x_combined, y_combined;
+    if (!need_combine) {
+      x_combined = *x;
+      y_combined = *y;
+    } else {
+      x_combined = is_fold_init_dims_x ? FoldInitDims(*x)
```

Comment:
Unfortunately changing to a version with a ternary operator in definition of "x_combined" and "y_combined" caused tests failure(something was wrong with dx and dy output gradients), so I will keep that version unless you have any idea why have it failed?

Reply:
I don't think I understand your question, my proposition is to get rid of `if else` blocks, just `if` should be enough

Reply:
Yes, I have tried that. I have replaced my code with:
```
Tensor x_combined = *x, y_combined = *y;
if(need_combine) {
 x_combined = is_fold_init_dims_x ? FoldOuterDims(*x) : FoldFirstAndLastDims<T>(dev_ctx, x); 
 y_combined = is_fold_init_dims_y ? FoldOuterDims(*y) : FoldFirstAndLastDims<T>(dev_ctx, y);
}
```
But this code was causing errors. Probably because of the way that operator= or constructor are implemented for Tensor class.

Reply:
OK